source/trace_event.c

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/clock.h>
+#include <aws/common/logging.h>
+#include <aws/common/trace_event.h>
+
+struct aws_trace_event *trace_event;
+uint64_t listener_id;
+uint64_t start_time;
+
+/*
+ * Private API
+ */
+
+/* Add trace event meta data to JSON object when the message bus allows it */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)msg;
+    cJSON *event = cJSON_CreateObject();
+
+    cJSON_AddItemToArray(trace_event->event_array, event);
+    /* add options for args later on, will initially be empty for now */
+    cJSON_AddItemToObject(event, "args", cJSON_CreateObject());
+
+    /* add more options later on */
+    cJSON_AddStringToObject(event, "cat", trace_event_data->category);
+    cJSON_AddStringToObject(event, "name", trace_event_data->name);
+    /* Fix for buffer overflow issue using char phase reference */
+    char ph[2];
+    strncat(ph, &(trace_event_data->phase), 1);
+    cJSON_AddStringToObject(event, "ph", ph);
+    cJSON_AddNumberToObject(event, "pid", trace_event_data->process_id);
+    cJSON_AddNumberToObject(event, "tid", trace_event_data->thread_id);
+    cJSON_AddNumberToObject(event, "ts", trace_event_data->timestamp);
+}
+
+/* Destructor of trace event metadata sent through the bus */
+void aws_trace_event_destroy(void *payload) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)payload;
+    aws_mem_release(trace_event->allocator, trace_event_data->name);
+    aws_mem_release(trace_event->allocator, trace_event_data->category);
+}
+
+/*
+ * Public API
+ */
+
+/* starts the message bus and initializes the JSON root, and the event array for storing metadata */
+int aws_trace_event_init(uint64_t address, struct aws_allocator *allocator) {
+    trace_event = aws_mem_acquire(allocator, sizeof(struct aws_trace_event));
+    trace_event->allocator = allocator;
+
+    trace_event->root = cJSON_CreateObject();
+    trace_event->event_array = cJSON_CreateArray();
+    cJSON_AddItemToObject(trace_event->root, "traceEvents", trace_event->event_array);
+
+    aws_high_res_clock_get_ticks(&start_time);
+    listener_id = address;
+
+    if (allocator == NULL || trace_event->root == NULL || trace_event->event_array == NULL) {
+        return AWS_OP_ERR;
```

Comment:
it's really hard to do great error handling. If you do 3 things, but the first 2 worked, you need to clean up those first 2 before returning error.

maybe just explode if something goes wrong here with the the CHECK() macro